graphix/gflow.py

```diff
@@ -12,228 +12,123 @@
-            vec.data[i_row] = 1
-            vec_add = adj_mat_row_reduced.data[:, node_order_list.index(node)]
-            vec += vec_add
-        elif meas_planes[node] == Plane.YZ:
-            vec.data = adj_mat_row_reduced.data[:, node_order_list.index(node)].reshape(vec.data.shape)
-        b.data[:, i_row] = vec.data
-    adj_mat, b, _, col_permutation = adj_mat.forward_eliminate(b)
-    x, kernels = adj_mat.backward_substitute(b)
-
-    corrected_nodes = set()
-    for i_row in range(len(node_order_row)):
-        non_out_node = node_order_row[i_row]
-        x_col = x[:, i_row]
-        if 0 in x_col.shape or x_col[0] == sp.nan:  # no solution
-            continue
-        if mode == "single":
-            sol_list = [x_col[i].subs(zip(kernels, [sp.false] * len(kernels))) for i in range(len(x_col))]
-            sol = np.array(sol_list)
-            sol_index = sol.nonzero()[0]
-            g[non_out_node] = {node_order_col[col_permutation.index(i)] for i in sol_index}
-            if meas_planes[non_out_node] in {Plane.XZ, Plane.YZ}:
-                g[non_out_node] |= {non_out_node}
-
-        elif mode == "all":
-            g[non_out_node] = set()
-            binary_combinations = product([0, 1], repeat=len(kernels))
-            for binary_combination in binary_combinations:
-                sol_list = [x_col[i].subs(zip(kernels, binary_combination)) for i in range(len(x_col))]
-                sol = np.array(sol_list)
-                sol_index = sol.nonzero()[0]
-                g_i = {node_order_col[col_permutation.index(i)] for i in sol_index}
-                if meas_planes[non_out_node] in {Plane.XZ, Plane.YZ}:
-                    g_i |= {non_out_node}
-
-                g[non_out_node] |= {frozenset(g_i)}
-
-        elif mode == "abstract":
-            g[non_out_node] = {}
-            for i in range(len(x_col)):
-                node = node_order_col[col_permutation.index(i)]
-                g[non_out_node][node] = x_col[i]
-            if meas_planes[non_out_node] in {Plane.XZ, Plane.YZ}:
-                g[non_out_node][non_out_node] = sp.true
-
-        l_k[non_out_node] = k
-        corrected_nodes |= {non_out_node}
-
-    if len(corrected_nodes) == 0:
-        if oset == nodes:
-            return g, l_k
-        return None, None
-    return gflowaux(
-        graph,
-        iset,
-        oset | corrected_nodes,
-        meas_planes,
-        k + 1,
-        l_k,
-        g,
-        mode=mode,
-    )
+    d = max(l_k.values())
+    layers: dict[int, set[int]] = {k: set() for k in range(d + 1)}
+    for i, val in l_k.items():
+        layers[val].add(i)
+    return d, layers
 
 
 def find_flow(
-    graph: nx.Graph,
-    iset: set[int],
-    oset: set[int],
-    meas_planes: dict[int, Plane] | None = None,
-) -> tuple[dict[int, set[int]], dict[int, int]]:
+    graph: nx.Graph[int],
+    iset: AbstractSet[int],
+    oset: AbstractSet[int],
+    meas: Mapping[int, Measurement] | None = None,
```

Comment:
The former type for `meas_planes` (`dict[int, Plane]`) expresses precisely what is required to compute a flow. I think we should keep it, in its abstract form `Mapping[int, Plane]`, rather than requiring unnecessary information. The function `generate_from_graph` will be responsible for converting `Measurement` to `Plane` before calling `find_flow`. The name `meas_planes` seemed appropriate but if we want something shorter, we could use just `planes` (although `meas` on its own would be misleading).